### PR TITLE
cpu/cc26xx_cc13xx: add AON_BATMON saul driver 

### PR DESCRIPTION
**Pull request by jeandudey**
### Contribution description

Add `AON_BATMON` saul driver. The CC26xx/CC13xx MCU family provides an Always-On Battery Monitor peripheral which measures the internal chip temperature and the supply voltage. The peripheral also offers events for the upper and lower voltage limits, although I think this can be done when we have something like #14182 .

That said, the battery montior has some inaccuracies on the temperature measurements.

<details>
<summary>make -C examples/saul flash term BOARD=cc1350-launchpad</summary>

```
2021-02-19 23:33:08,999 # Welcome to RIOT!
2021-02-19 23:33:09,000 # 
2021-02-19 23:33:09,005 # Type `help` for help, type `saul` to see all SAUL devices
2021-02-19 23:33:09,005 # 
> saul
2021-02-19 23:33:10,307 # saul
2021-02-19 23:33:10,308 # ID	Class		Name
2021-02-19 23:33:10,311 # #0	ACT_SWITCH	LED(red)
2021-02-19 23:33:10,313 # #1	ACT_SWITCH	LED(green)
2021-02-19 23:33:10,315 # #2	SENSE_BTN	Button(BTN-1)
2021-02-19 23:33:10,318 # #3	SENSE_BTN	Button(BTN-2)
2021-02-19 23:33:10,320 # #4	SENSE_VOLTAGE	CC26XX_CC13XX_BAT_V
2021-02-19 23:33:10,324 # #5	SENSE_TEMP	CC26XX_CC13XX_BAT_T
> saul read 4
2021-02-19 23:33:16,588 # saul read 4
2021-02-19 23:33:16,593 # Reading from #4 (CC26XX_CC13XX_BAT_V|SENSE_VOLTAGE)
2021-02-19 23:33:16,595 # Data:	           1152 mV
> saul read 5
2021-02-19 23:33:18,914 # saul read 5
2021-02-19 23:33:18,919 # Reading from #5 (CC26XX_CC13XX_BAT_T|SENSE_TEMP)
2021-02-19 23:33:18,921 # Data:	            -11 °C
> 
```
</details>

The values both for voltage and temperature seem to be out of range, however launchpads aren't connected to a battery (at least for the voltage), tomorrow I will test with a cc1312 custom board to see any differences. It might also be that the cc13x0 need the `SetupTrimDevice` function implemented as it touches a register related to it.

### Testing procedure

- `make -C examples/saul flash term BOARD=cc1312-launchpad`
- `make -C examples/saul flash term BOARD=cc1350-launchpad`
- `make -C examples/saul flash term BOARD=cc1352-launchpad`
- `make -C examples/saul flash term BOARD=cc1352p-launchpad`

### Issues/PRs references

### CODE REVIEW

File: drivers/saul/init_devs/auto_init_saul_cc26xx_cc13xx_batmon.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ */
+
+/**
+ * @ingroup     sys_auto_init_saul
+ * @{
+ *
+ * @file
+ * @brief       Auto initialization of internal temperature and battery voltage
+ *              sensor directly mapped to SAUL reg
```
Comment (jeandudey) [inline drivers/saul/init_devs/auto_init_saul_cc26xx_cc13xx_batmon.c:16]: Oops, to fix tomorrow :-)

Comment (jeandudey) [reply drivers/saul/init_devs/auto_init_saul_cc26xx_cc13xx_batmon.c:16]: Nevermind, it's fine to be in that group and description matches.

**Comment by fjmolinas** [comment]
@jeandudey where you able to test on your custom Board? -11C seems a little bit odd..

**Comment by jeandudey** [comment]
@fjmolinas 

Seems to be that it happens only on `cc13x0` devices, using a cc1312 (which trims some registers at initialization) gives more accurate values. If anyone ha s a `cc2650` around it should work too by the same logic (any test would be appreciated :P).

<details>
<summary>Test on custom board</summary>

```
jeandudey@jean ~/D/RIOT> make -C examples/saul BOARD=cc1312-launchpad
make: Entering directory '/home/jeandudey/Dev/RIOT/examples/saul'
Building application "saul_example" for "cc1312-launchpad" with MCU "cc26x2_cc13x2".

"make" -C /home/jeandudey/Dev/RIOT/boards/cc1312-launchpad
"make" -C /home/jeandudey/Dev/RIOT/core
"make" -C /home/jeandudey/Dev/RIOT/cpu/cc26x2_cc13x2
"make" -C /home/jeandudey/Dev/RIOT/cpu/cc26x2_cc13x2/periph
"make" -C /home/jeandudey/Dev/RIOT/cpu/cc26xx_cc13xx
"make" -C /home/jeandudey/Dev/RIOT/cpu/cc26xx_cc13xx/batmon
"make" -C /home/jeandudey/Dev/RIOT/cpu/cc26xx_cc13xx/periph
"make" -C /home/jeandudey/Dev/RIOT/cpu/cortexm_common
"make" -C /home/jeandudey/Dev/RIOT/cpu/cortexm_common/periph
"make" -C /home/jeandudey/Dev/RIOT/drivers
"make" -C /home/jeandudey/Dev/RIOT/drivers/periph_common
"make" -C /home/jeandudey/Dev/RIOT/drivers/saul
"make" -C /home/jeandudey/Dev/RIOT/drivers/saul/init_devs
"make" -C /home/jeandudey/Dev/RIOT/sys
"make" -C /home/jeandudey/Dev/RIOT/sys/auto_init
"make" -C /home/jeandudey/Dev/RIOT/sys/fmt
"make" -C /home/jeandudey/Dev/RIOT/sys/isrpipe
"make" -C /home/jeandudey/Dev/RIOT/sys/malloc_thread_safe
"make" -C /home/jeandudey/Dev/RIOT/sys/newlib_syscalls_default
"make" -C /home/jeandudey/Dev/RIOT/sys/phydat
"make" -C /home/jeandudey/Dev/RIOT/sys/ps
"make" -C /home/jeandudey/Dev/RIOT/sys/saul_reg
"make" -C /home/jeandudey/Dev/RIOT/sys/shell
"make" -C /home/jeandudey/Dev/RIOT/sys/shell/commands
"make" -C /home/jeandudey/Dev/RIOT/sys/stdio_uart
"make" -C /home/jeandudey/Dev/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  17152	    188	   2436	  19776	   4d40	/home/jeandudey/Dev/RIOT/examples/saul/bin/cc1312-launchpad/saul_example.elf
make: Leaving directory '/home/jeandudey/Dev/RIOT/examples/saul'
jeandudey@jean ~/D/RIOT> ti-sbl-prog -p /dev/ttyUSB0 --baudrate 1500000 flash /home/jeandudey/Dev/RIOT/examples/saul/bin/cc1312-launchpad/saul_example.elf --write-erase
Error: Binary size is too large
jeandudey@jean ~/D/RIOT> ti-sbl-prog -p /dev/ttyUSB0 --baudrate 1500000 flash /home/jeandudey/Dev/RIOT/examples/saul/bin/cc1312-launchpad/saul_example.bin --write-erase
   Sectors erased
   Transfers finished
jeandudey@jean ~/D/RIOT> make -C examples/saul BOARD=cc1312-launchpad term PORT=/dev/ttyUSB0
make: Entering directory '/home/jeandudey/Dev/RIOT/examples/saul'
/home/jeandudey/Dev/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyUSB0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-02-23 11:45:56,863 # Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
2021-02-23 11:46:00,576 # main(): This is RIOT! (Version: 2021.04-devel-694-g317e3-2021_02_16-cc26xx_cc13xx-saul-batmon)
2021-02-23 11:46:00,577 # Welcome to RIOT!
2021-02-23 11:46:00,577 # 
2021-02-23 11:46:00,578 # Type `help` for help, type `saul` to see all SAUL devices
2021-02-23 11:46:00,579 # 
> help
2021-02-23 11:46:02,880 # help
2021-02-23 11:46:02,881 # Command              Description
2021-02-23 11:46:02,882 # ---------------------------------------
2021-02-23 11:46:02,896 # reboot               Reboot the node
2021-02-23 11:46:02,897 # version              Prints current RIOT_VERSION
2021-02-23 11:46:02,899 # pm                   interact with layered PM subsystem
2021-02-23 11:46:02,900 # ps                   Prints information about running threads.
2021-02-23 11:46:02,913 # saul                 interact with sensors and actuators using SAUL
> saul
2021-02-23 11:46:04,464 # saul
2021-02-23 11:46:04,464 # ID	Class		Name
2021-02-23 11:46:04,465 # #0	ACT_SWITCH	LED(red)
2021-02-23 11:46:04,466 # #1	ACT_SWITCH	LED(green)
2021-02-23 11:46:04,466 # #2	SENSE_BTN	Button(BTN-1)
2021-02-23 11:46:04,480 # #3	SENSE_BTN	Button(BTN-2)
2021-02-23 11:46:04,481 # #4	SENSE_VOLTAGE	CC26XX_CC13XX_BAT_V
2021-02-23 11:46:04,482 # #5	SENSE_TEMP	CC26XX_CC13XX_BAT_T
> saul read 4
2021-02-23 11:46:07,408 # saul read 4
2021-02-23 11:46:07,409 # Reading from #4 (CC26XX_CC13XX_BAT_V|SENSE_VOLTAGE)
2021-02-23 11:46:07,410 # Data:	           1027 mV
> saul read 5
2021-02-23 11:46:11,136 # saul read 5
2021-02-23 11:46:11,138 # Reading from #5 (CC26XX_CC13XX_BAT_T|SENSE_TEMP)
2021-02-23 11:46:11,139 # Data:	              7 °C
> 
```
</details>

Gives 7 °C and 1027 mV from the supply (which seems to be an odd value too, but it doesn't use a battery)

**Comment by benpicco** [comment]
I ran this on the `cc1352p-launchpad` and I also get

```
2021-02-25 21:28:04,520 # Reading from #5 (CC26XX_CC13XX_BAT_T|SENSE_TEMP)
2021-02-25 21:28:04,522 # Data:	              7 °C
2021-02-25 21:29:56,547 #  saul read 4
2021-02-25 21:29:56,552 # Reading from #4 (CC26XX_CC13XX_BAT_V|SENSE_VOLTAGE)
2021-02-25 21:29:56,554 # Data:	           1027 mV
```

And it's really not *that* cold in my room.
Code wise this looks good though.

**Comment by jeandudey** [comment]
I have re-readed the Technial Reference Manual and while it still doesn't give in how much time the battery monitor takes a measurement it specifies one register to check if it's updated (it does say that the batmon ADC accumulates the information through time and updates the values depending on that). I will leave this night the launchpad reading values (and checking for that flag, to clear it and re-check) to see if the values changes, after all I don't think it's a coincidence that a cc1312 and cc1352p show the same value at a fresh start up of the device.

**Comment by jeandudey** [comment]
No luck, same values all the time:

```2021-02-27 16:29:11,692 # ##########################
2021-02-27 16:29:12,655 # 
2021-02-27 16:29:12,657 # Dev: LED(red)	Type: ACT_SWITCH
2021-02-27 16:29:12,659 # Data:	              0 
2021-02-27 16:29:12,659 # 
2021-02-27 16:29:12,662 # Dev: LED(green)	Type: ACT_SWITCH
2021-02-27 16:29:12,664 # Data:	              0 
2021-02-27 16:29:12,664 # 
2021-02-27 16:29:12,667 # Dev: Button(BTN-1)	Type: SENSE_BTN
2021-02-27 16:29:12,669 # Data:	              0 
2021-02-27 16:29:12,669 # 
2021-02-27 16:29:12,672 # Dev: Button(BTN-2)	Type: SENSE_BTN
2021-02-27 16:29:12,674 # Data:	              0 
2021-02-27 16:29:12,674 # 
2021-02-27 16:29:12,678 # Dev: CC26XX_CC13XX_BAT_V	Type: SENSE_VOLTAGE
2021-02-27 16:29:12,680 # Data:	           1152 mV
2021-02-27 16:29:12,680 # 
2021-02-27 16:29:12,684 # Dev: CC26XX_CC13XX_BAT_T	Type: SENSE_TEMP
2021-02-27 16:29:12,686 # Data:	            -11 °C
2021-02-27 16:29:12,687 # 
2021-02-27 16:29:12,689 # ##########################
```

And it's not that could :P

I don't know what I'm doing wrong, will leave open the PR until I find a solution or I'll drop it since it doesn't provide any valuable information.

**Comment by jeandudey** [comment]
Closing as I couldn't get a reasonable value from the device peripheral, almost no documentation online other than the examples, however this code doesn't do anything different when taking the readings and enabling the peripheral measurements.

**Pull request closed without merging**